### ISSUE TITLE
Getting state in RoutesMap fromPath toPath?

### ISSUE THREAD

**Issue by waynebloss** [faceyspacey/redux-first-router#220]
**TLDR:** I had to use a global var to store a mapping that I want to use in [`fromPath/toPath`](https://github.com/faceyspacey/redux-first-router/blob/master/docs/connectRoutes.md#routesmap), instead of using my redux state.

I have a multi-tenant app similar to the one mentioned in #202 but instead of `'/company/:companyID/posts'`, my path patterns are just `'/:orgId/'`, `':orgId/posts/'` and so on...

Also instead of `orgId` being a number, it's a slug string that I convert to and from a number using the `fromPath` and `toPath` options of each RoutesMap entry [as seen in the docs](https://github.com/faceyspacey/redux-first-router/blob/master/docs/connectRoutes.md#routesmap).

Here is my code for an organizations default page at `'/:orgId/'`. With this code, when I go to `http://www.mysite.test/hello-inc`, my location `payload.orgId` will not contain `'hello-inc'` but instead 999, the numeric id of *Hello, Inc.*.

```js
  fromPath: (slug, key) => {
    if (key === 'orgId') {
      // Return an id number for the given slug, from the users authorized organizations.
      return window.orgIdBySlug[slug];
    }
    return slug;
  },
  toPath: (value, key) => {
    if (key === 'orgId') {
      // Return a slug for the given id number, from the users authorized organizations.
      return window.orgSlugById[value];
    }
    return value;
  },
```

My only problem, as you can see, is that I had to resort to using a global variable to hold the information that I needed to do this conversion. (UPDATE: It doesn't have to be a global var, but wherever the mapping is stored, it cannot be in the redux store. See [this comment](https://github.com/faceyspacey/redux-first-router/issues/220#issuecomment-372197573))

The problem is that fromPath gets called before `createStore` even returns, even if I pass `initialDispatch: false` to the [connectRoutes options](https://github.com/faceyspacey/redux-first-router/blob/master/docs/connectRoutes.md#options). So, I couldn't even use a global reference to my state store unless I make the first call to `fromPath` fail or return a dummy value.

Any suggestions?

The reason I don't want to use a global var if I don't have to is because there's a bit of work involved in keeping it up to date and in sync with the redux state which could be changed across tabs and auto rehydrated. I also had to add side effects to some of my reducers, in order to update these global/non-redux-state vars.

**Comment by waynebloss** [comment]
Besides this one little problem, it was really easy to do private, multi-tenant, client-sided routing with this kit.

My users will be involved in only a handful of organizations each, so I basically get that list of org `{id, name, slug}` upon login and save it on the client so I can properly route every logged in user to their organizational pages. So, this solution works really great.

**Comment by waynebloss** [comment]
Correction: I didn't have to actually use a global variable here. It was just easier to do that in my alpha code. Since then, I've organized this code into a new module with a private variable that holds the map.

I still have to store a variable outside of my state and sync it with my state to implement this, but at least the global var is gone.

**Comment by faceyspacey** [comment]
In the real Rudy, the `extra` argument and all of your options + routes will be passed to these callbacks.

For now, you have options to use the `store`:

- define those routes in the same file where `store` is closed over so you have access to them in the functions. 
- `export getStore = () => store` from your `configureStore` file and it will be available to get IDs out of.

The one caveat in both new and old Rudy is that the store won't exist for the first route! So you have to work around that. But obviously if you're doing this approach it's because you have sent the hash of slugs/ids to the browser and have them already (or somehow don't need them on the first route), and your concern is just later having an *up to date* "reactive" version without having to resolve to workarounds such as `store.subscribe(updateHashes)`. 

So basically the best you can do in new and old Rudy is, 

```js
to/fromPath: () => {
   if (!store) // get from hash that doesn't exist in store yet
   else store.getState().hash[slugOrId] // etc
}
```

Again that's because the store won't exist no matter what for `initialLocationReducerState`.

in real Rudy, you can inject the initial hash as `options.extra` and it will be passed to these functions. But you still gotta do if/else. Well, actually you could do this:

```js
options.extra = {
   getHash: store ? store.getState().hash : initialHash
}
```

then something like:

```js
to/fromPath: (slug, key, opts) => opts.extra.getHash()[slug]
```

So the dream of the least code you're after here will be possible :)

I wouldn't worry about it for now. It sounds like you have a working solution. Hopefully some info here can help you upgrade it a bit.
